### DISCUSSION TITLE
Foundry StableCoin unit test redeemCollateral fails a reentrancy call

### DISCUSSION THREAD

**Discussion by NeoRusi** [Cyfrin/foundry-full-course-cu#840] [Q&A]
hey guys here is my redeemCollateral
```solidity
function redeemCollateral(address tokenCollateralAddress, uint256 amountCollateral)
        public
        moreThanZero(amountCollateral)
        nonReentrant
    {
        _redeemCollateral(tokenCollateralAddress, amountCollateral, msg.sender, msg.sender);
        _revertIfHealthFactorIsBroken(msg.sender);
    }
```
I was writing unit test, challenging myself, and while writing the I found a bug in the code(which I found but I don't understand currently)
```
Running 3 tests for test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: ReentrancyGuard: reentrant call] testRedeemCollateral() (gas: 110623)
[PASS] testRedeemCollateralIsNonReentrant() (gas: 3451)
[PASS] testRedeemCollateralRevertsIfAmountCollateralLessThanZero() (gas: 41464)
Test result: FAILED. 2 passed; 1 failed; 0 skipped; finished in 7.59ms
 
Ran 1 test suites: 2 tests passed, 1 failed, 0 skipped (3 total tests)

Failing tests:
Encountered 1 failing test in test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: ReentrancyGuard: reentrant call] testRedeemCollateral() (gas: 110623)
```
here is my unit test
```solidity
    function testRedeemCollateral() public depositCollateral {
        vm.startPrank(USER);
        dsce.redeemCollateral(weth, AMOUNT_COLLATERAL);
        uint256 userBalance = ERC20Mock(weth).balanceOf(USER);
        assertEq(userBalance, AMOUNT_COLLATERAL);
        vm.stopPrank();
    }
```
Maybe someone now how to deal with this

**Comment by theirrationalone** [comment]
Please provide full traces of that test, copy paste all the output of `forge test testname -vvvv` and paste here. Then after, we may could help you. 😅

**Comment by theirrationalone** [comment]
Or check if you have `nonreentrant` modifier with `_redeemCollateral` function in your dscengine.sol?? 

If you have, please remove that !

**Reply by NeoRusi** [reply]
yeh already understood but doesn't the redeem collateral function should be nonreentrant cause it moves tokens

**Reply by theirrationalone** [reply]
Of course `redeemcollateral` must has a `nonreentrant` modifier and it has so, cause as you said, it moves tokens. Haven't you noticed, `redeemcollateral` is a public function has a `nonreentrant` modifier, calling `_redeemcollateral` internal function prepended with an underscore. That's why we can't put `nonreentrant` modifier again with `_redeemcollateral` internal function. Therefore it already has that modifier with `redeemcollateral` public function and we can't call an internal function from outside of a contract.

So we can't put `nonreentrant` more than one time.
